---
title: "Lab module 3 – Part 1"
output:
  html_document:
    toc: true
    toc_float: true
    collapsed: false
    number_sections: false
    toc_depth: 3
    #code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE,warning=FALSE, cache=T)
```

```{r klippy, echo=FALSE, include=TRUE}
klippy::klippy()
```

*Lab aims and steps:*

*1.  Estimate SIR models for covid-19 variants*
    
*2.  Classify covid-19 variants and predict Ro for emerging variant*
    
<br>

**Due:** Nothing at the end of this lab.  Be sure to save all your code and any questions or tasks in <mark style="color:blue">*blue text.*</mark>

For the next two labs we will focus on using tools of machine learning to describe and predict the spread of covid-19.  In the first week we will focus on generating SIR models and using classification methods to make predictions about an emerging variant.

<br>

### 1. Estimate SIR models for covid-19 variants

To generate SIR models, first we need data on infections.  We can get this from a wonderful resource called [GISAID](https://gisaid.org/), which tracks flu and covid cases.  It has the number of infections by country and variant over time.  Variant identification is based on genome sequencing of the virus.  The data is available by permission only, so please do not share it with anyone outside of this class.

On Canvas, I have provided you with the United States infection data for 4 of the variants of concern: Alpha, Beta, Delta, and Omicron.  For each of these variants, you will develop an SIR model and from it estimate the basic reproductive number, Ro.

You should work through ALL of the code in this section for each variant, one at a time (the example code below is for the alpha variant).

First, you need to upload (or download, then upload) the required packages and the data into Rstudio, then do some housecleaning.

```{r}
# Required packages
require(deSolve)
require(ggplot2)

# upload data
pdata <- read.csv("C:/Users/17163/Documents/DSEE/Labs/Covid/Alpha.csv",header=T)
pdata <- data.frame(as.numeric(pdata$week),as.numeric(pdata$count))
names(pdata) <- c("t","I")

# do some housecleaning
pdata$I[is.na(pdata$I)] <- 0
t <- pdata$t
I <- pdata$I
cases <- pdata$I
```

The dataset is composed of two variables t (time), or the week of the pandemic, and I, the number of infected individuals.  Week 1 started on Dec. 29, 2019 and week 147 is the week of October 16, 2022.  Note, I only included data for the Omicron variant through March 6, 2022, which is before the more transmissible subvariants (BA.2-BA.5) began to dominate in the US. 

The first thing we need to do to produce the SIR model is to make a function that calculates the change in each category (S, I, and R) per unit of time using the ordinary differential equations.

```{r}
# Function for calculating S, I, R values
sir <- function(t,x,parms){
  S <- x[1]
  I <- x[2]
  beta <- parms[1]
  gamma <- parms[2]
         dS <- -(beta*S*I)
         dI <- beta*S*I - gamma*I
         res <- c(dS,dI)
         list(res)
}
```

Note that we are only actually calculating S and I.  If we want R, we can get the numbers easily by multiplying I*gamma.

Next, we need to find the values of beta and gamma.  Beta (transmission rate) and gamma (recovery rate) are the parameters that dictate the rate at which individuals transition from S to I (beta) and I to R (gamma).  To determine the values that best fit the data, we will use the sum of squared error (residuals) values for the number of infected individuals (I).  Basically we are calculating how much our model values differ from the actual values.

```{r}
# function for calculating sum of squared error for optimization
sse.sir <- function(parms0,data) {
  data <- data
  t <- data[,1]
  cases <- data[,2]
  beta <- exp(parms0[1])
  gamma <- exp(parms0[2])
  S0 <- exp(parms0[3])
  I0 <- exp(parms0[4])
  parms <- c(beta, gamma)
  out <- ode(y=c(S0,I0), t, sir, parms, hmax=1/120)
  print(sum((out[,3]-cases)^2)) # equation for sum of squared error 
}
```


The next step is to find the beta and gamma values (along with initial S and I values) that minimize the sum of squared error.  This is called optimization.  We aren’t going to go into the weeds on this, but there are many algorithms for optimization and we are using the [Nelder and Mead method](https://en.wikipedia.org/wiki/Nelder%E2%80%93Mead_method) within the <mark style="background-color:grey">optim()</mark> function (use ?optim for more information). 

To optimize our parameter values (beta, gamma, S0, and I0) we need to give the algorithm initial values.  Unfortunately, most of the algorithms are pretty sensitive to the starting values, so I have provided you with unique initial values for each variants.

When you run the optimization line of code given below R will spit out the sum of squared error values for all of the parameters it is trying.  You should notice that the values decrease in size.

The last two lines of code in this set give you the parameter estimates and an estimate of Ro.  Here Ro is beta*S0/gamma.  It includes initial S (S0) because the ordinary differential equations we used here did not include the population size.  Getting the correct pop. size estimate value for a specific variant and group of susceptible individuals is hard, so we use a version of the equations without population size.

```{r}
# Initial parameter values for optimization
# alpha
parms0 <- c(-9,-9,12,-2.6)

# Omicron
#parms0 <- c(-9.5,-9,12.5,-2.6)

# Delta
#parms0 <- c(-9,-6,12,-2.6)

# Beta
#parms0 <- c(-10,-10,8,-2.6)

# optimization
fit1 <- optim(parms0, sse.sir,data = pdata)

# parameter values (beta, gamma, S0, I0)
exp(fit1$par)

# Estimate of Ro
exp(fit1$par[1])*exp(fit1$par[3])/exp(fit1$par[2])
```

The next code block produces a plot of the observed and optimized model values for the number of infected individuals.

```{r}
#produce model predicted values to plot
mod.pred <- data.frame(ode(c(exp(fit1$par[3]),exp(fit1$par[4])),t,
                           sir,c(exp(fit1$par[1]),exp(fit1$par[2])),hmax=1/120))
names(mod.pred) <- c("t","S","I")

#plot observed and modeled predictions
(Iplot <- ggplot(pdata, aes (x = t, y = I)) +
  geom_point(alpha = 0.5, size = 3) + 
  geom_line(data = mod.pred, aes(x = t, y = I), col = "red", size = 1.5) + 
  theme_bw() +
  ylab("Infected\n") +                             
  xlab("\nWeek")  +
  theme(axis.text.x = element_text(size = 14),
        axis.text.y = element_text(size = 14),
        axis.title = element_text(size = 16, face = "plain"),                        
        panel.grid = element_blank(),                                                 
        plot.margin = unit(c(1,1,1,1), units = , "cm")))
```

<mark style="color:blue">*After you work through all of the above code for each of the four variants, I want you to alter the code so that you can produce a SINGLE publication quality plot that has the observed and optimized model infection values for all four variants.*</mark>

Do the Ro values for any of the variants seem too high to low?  If you are unsure, Google the current best estimates of the Ro values for each of the variants.  What could be possible causes for these inaccurate estimates?  Think both about issues with optimization (hint: look at the plots) and violation of SIR assumptions.

<br>

### 2. Classify covid-19 variants and predict Ro for emerging variant

Next up, we are going to use sars-cov-2 spike protein sequences to classify variants and, in conjunction with results from part 1, estimate the Ro for an emerging variant.  The spike protein plays an important role in binding of the sars-cov-2 virus to human cells, and influences the transmission and virulence of the disease (eg, [Huang et al 2020](https://www.nature.com/articles/s41401-020-0485-4)).  The protein sequences come from NCBI, which has a decent [dashboard](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/) for downloading covid related data.

To classify the emerging variant, we are going to use [principle component analysis](https://en.wikipedia.org/wiki/Principal_component_analysis) (PCA) to reduce the dimensionality of the data (protein sequences are quite complex), then use the k-nearest neighbor method (remember back to the lecture from lab last week) to determine which of the [variants of concern](https://www.who.int/en/activities/tracking-SARS-CoV-2-variants/) the emerging variant is most similar to and use the variant of concern Ro as a first step prediction of the emerging variant’s basic reproductive number.

The k-nearest neighbor method uses a training dataset (called sequences_train on Canvas), in this case, one where we know the variants (and their Ro values).  We use the training dataset to classify the emerging, or unknown, variant (found in the test dataset called sequences_test on Canvas).  The training dataset is composed of 20 sequences for each of four variants of concern: Alpha, Beta, Delta, and Omicron.  The test dataset is composed of 5 sequences.

Let’s start by getting the needed packages and data into R.  

```{r}
# required packages
require(ape)
require(kmer)
require(class)
require(ggfortify)

# training data, the data come in what is called fasta format
train <- read.FASTA("/Users/17163/Documents/DSEE/Labs/Covid/sequences_train.txt",type="AA")

# training data IDs (variant)
vID <- c(rep("Alpha",10),rep("Delta",10),rep("Beta",10),rep("Omicron",10),
               rep("Alpha",10),rep("Beta",10),rep("Delta",10),rep("Omicron",10))

# test data (unknowns)
test <- read.FASTA("/Users/17163/Documents/DSEE/Labs/Covid/sequences_test.txt",type="AA")

# combining training and test data
comb <- c(train,test)

# adding test data unknown variant
vID_t <- c(vID, rep("U",5))
```

The first step in our classification procedure is to produce a distance matrix.  This is basically a measure of similarity between the sequences.  The higher the number the more similar the sequences.  But then we have an 85x85 matrix of distance values.  So, we use principal component analysis to reduce the complexity down to two variables using PCA.  PCA maximizes the amount of variation that can be explained in a smaller number of variables, which in this case is two.  PCA also allows us to plot the data more easily.

```{r}
# calculate distance matrix
mat_t <- kdistance(comb,k=2, method = "edgar")

# reformat and calculate principle components
out_t <- kmeans(mat_t,4)
re_mat_t <- as.matrix(mat_t)
plot <- autoplot(out_t, mat_t, frame = TRUE)
```

Before we use these data to help classify the unknown emerging variant sequences, let’s plot the results of the PCA.

```{r}
# make data frame of PCs and Variant IDs
plot_data <- data.frame(plot$data$PC1,plot$data$PC2,vID_t)
names(plot_data) <- c("dem1","dem2","Variant")

# colors for plot
colors <- c("red","green","blue","purple","orange")

# plot data
(d_plot <- ggplot(plot_data, aes (x = dem1, y = dem2, color=Variant)) +
    geom_point(alpha = 0.5, size = 3) + 
    scale_color_manual(values = colors) +
    theme_bw() +
    ylab("PC2\n") +                             
    xlab("\nPC1")  +
    theme(axis.text.x = element_text(size = 14),
          axis.text.y = element_text(size = 14),
          axis.title = element_text(size = 16, face = "plain"),                        
          panel.grid = element_blank(),                                                 
          plot.margin = unit(c(1,1,1,1), units = , "cm")))
```

The plot has the principal component variables 1 and 2 on the x and y axes and variants IDs by color.  You can see that the PCs do a pretty good (not perfect!) job of grouping the sequences for variants together.  For example, all of the Omicron sequences are clumped nicely at the top in purple. 

The last step is to use the k-nearest neighbor method to determine which variant of concern the uknowns are most similar to.  More specifically, the code below finds the five nearest neighbors for each of the unknowns and outputs the variant it would assign each emerging sequence to.  It outputs them simply as a list.  

```{r}
# finds the 20 (k=20) nearest neighbors for the unknowns and assigns  
# them to a group (variant)
knn(plot_data[1:80,1:2],plot_data[81:85,1:2],cl=plot_data[1:80,3],k=20, prob = T)
```

You may have also noticed that I specified an argument called “prob” to be true in the above code.  This give the proportion of nearest neighbors within the class or group to which the algorithm assigns the unknowns (emerging variant sequences). 

I developed a function (given below) that uses this proportion of nearest neighbors to estimate the Ro with a weighted mean based on the proportions given.

```{r}
# function for finding k nearest neighbors
# and calculating the proportion for each class (variant)

# make data frame of estimated Ro by variant (order is Alpha, Delta, Beta, Omicron)
Ro <- c(1.8,5.8,1,8.5)
Variant <- unique(vID_t)[-length(unique(vID_t))]
Ro_d <- data.frame(Ro,Variant)

# x is the PCs, y is vID_t (variant labels), z is the number of neighbors considered, zz is Ro_d(Ro for each variant)
knnd <- function(x,y,z,zz) {
  d1 <- as.matrix(dist(x[,1:ncol(x)])) # generate distance matrix
  d2 <- d1[y=="U",y!="U"]              # subset only unknowns
  o <- y[apply(d2,1,order)[1:z,]]      # order by nearest neighbors
  o <- matrix(o,nrow=z)                # convert to matrix
  tt <- apply(o,2,table)               # determine proportion of z neighbors for each variant
  Ro_w<-numeric(ncol(o))
  var<-numeric(ncol(o))
  for (i in 1:ncol(o)){
    Ro_w[i]<-sum(zz$Ro[match(names(tt[[i]]),zz$Variant)]*tt[[i]])/z #calc weighted mean
    if (length(names(which.max(tt[[i]])))==0) {var[i]<-NA} else {var[i]<-names(which.max(tt[[i]]))} # determine most similar variant
  }
  print(var)  # print results
  print(Ro_w) 
}

# run function on your data
knnd(plot_data[1:85,1:2],vID_t,20,Ro_d)
```


<mark style="color:blue">*Write these results in your own words and be sure to include your best estimate for the Ro.*</mark>

<mark style="color:blue">*Together in lab, we will go through the process of finding and downloading a sequence from the NCBI website.  Then, you can repeat the process in part 2 to make predictions about another “unknown variant” sequence.*</mark>

<mark style="color:blue">*Be sure to save the principal component plot with your additional sequence plotted and an interpretation of your results.*</mark>
