<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Lab module 3 – Part 1</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">DSEE Labs</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="DSEE-Lab1.html">DSEE Lab #1</a>
</li>
<li>
  <a href="DSEE-Lab2.html">DSEE Lab #2</a>
</li>
<li>
  <a href="DSEE-Lab3.html">DSEE Lab #3</a>
</li>
<li>
  <a href="DSEE-Lab4.html">DSEE Lab #4</a>
</li>
<li>
  <a href="DSEE-Lab5and6.html">DSEE Lab #5 and 6</a>
</li>
<li>
  <a href="DSEE-Lab7.html">DSEE Lab #7</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Lab module 3 – Part 1</h1>

</div>


<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('left', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<p><em>Lab aims and steps:</em></p>
<p><em>1. Estimate SIR models for covid-19 variants</em></p>
<p><em>2. Classify covid-19 variants and predict Ro for emerging
variant</em></p>
<p><br></p>
<p><strong>Due:</strong> Nothing at the end of this lab. Be sure to save
all your code and any questions or tasks in
<mark style="color:blue"><em>blue text.</em></mark></p>
<p>For the next two labs we will focus on using tools of machine
learning to describe and predict the spread of covid-19. In the first
week we will focus on generating SIR models and using classification
methods to make predictions about an emerging variant.</p>
<p><br></p>
<div id="estimate-sir-models-for-covid-19-variants"
class="section level3">
<h3>1. Estimate SIR models for covid-19 variants</h3>
<p>To generate SIR models, first we need data on infections. We can get
this from a wonderful resource called <a
href="https://gisaid.org/">GISAID</a>, which tracks flu and covid cases.
It has the number of infections by country and variant over time.
Variant identification is based on genome sequencing of the virus. The
data is available by permission only, so please do not share it with
anyone outside of this class.</p>
<p>On Canvas, I have provided you with the United States infection data
for 4 of the variants of concern: Alpha, Beta, Delta, and Omicron. For
each of these variants, you will develop an SIR model and from it
estimate the basic reproductive number, Ro.</p>
<p>You should work through ALL of the code in this section for each
variant, one at a time (the example code below is for the alpha
variant).</p>
<p>First, you need to upload (or download, then upload) the required
packages and the data into Rstudio, then do some housecleaning.</p>
<pre class="r"><code># Required packages
require(deSolve)
require(ggplot2)

# upload data
pdata &lt;- read.csv(&quot;C:/Users/17163/Documents/DSEE/Labs/Covid/Alpha.csv&quot;,header=T)
pdata &lt;- data.frame(as.numeric(pdata$week),as.numeric(pdata$count))
names(pdata) &lt;- c(&quot;t&quot;,&quot;I&quot;)

# do some housecleaning
pdata$I[is.na(pdata$I)] &lt;- 0
t &lt;- pdata$t
I &lt;- pdata$I
cases &lt;- pdata$I</code></pre>
<p>The dataset is composed of two variables t (time), or the week of the
pandemic, and I, the number of infected individuals. Week 1 started on
Dec. 29, 2019 and week 147 is the week of October 16, 2022. Note, I only
included data for the Omicron variant through March 6, 2022, which is
before the more transmissible subvariants (BA.2-BA.5) began to dominate
in the US.</p>
<p>The first thing we need to do to produce the SIR model is to make a
function that calculates the change in each category (S, I, and R) per
unit of time using the ordinary differential equations.</p>
<pre class="r"><code># Function for calculating S, I, R values
sir &lt;- function(t,x,parms){
  S &lt;- x[1]
  I &lt;- x[2]
  beta &lt;- parms[1]
  gamma &lt;- parms[2]
         dS &lt;- -(beta*S*I)
         dI &lt;- beta*S*I - gamma*I
         res &lt;- c(dS,dI)
         list(res)
}</code></pre>
<p>Note that we are only actually calculating S and I. If we want R, we
can get the numbers easily by multiplying I*gamma.</p>
<p>Next, we need to find the values of beta and gamma. Beta
(transmission rate) and gamma (recovery rate) are the parameters that
dictate the rate at which individuals transition from S to I (beta) and
I to R (gamma). To determine the values that best fit the data, we will
use the sum of squared error (residuals) values for the number of
infected individuals (I). Basically we are calculating how much our
model values differ from the actual values.</p>
<pre class="r"><code># function for calculating sum of squared error for optimization
sse.sir &lt;- function(parms0,data) {
  data &lt;- data
  t &lt;- data[,1]
  cases &lt;- data[,2]
  beta &lt;- exp(parms0[1])
  gamma &lt;- exp(parms0[2])
  S0 &lt;- exp(parms0[3])
  I0 &lt;- exp(parms0[4])
  parms &lt;- c(beta, gamma)
  out &lt;- ode(y=c(S0,I0), t, sir, parms, hmax=1/120)
  print(sum((out[,3]-cases)^2)) # equation for sum of squared error 
}</code></pre>
<p>The next step is to find the beta and gamma values (along with
initial S and I values) that minimize the sum of squared error. This is
called optimization. We aren’t going to go into the weeds on this, but
there are many algorithms for optimization and we are using the <a
href="https://en.wikipedia.org/wiki/Nelder%E2%80%93Mead_method">Nelder
and Mead method</a> within the
<mark style="background-color:grey">optim()</mark> function (use ?optim
for more information).</p>
<p>To optimize our parameter values (beta, gamma, S0, and I0) we need to
give the algorithm initial values. Unfortunately, most of the algorithms
are pretty sensitive to the starting values, so I have provided you with
unique initial values for each variants.</p>
<p>When you run the optimization line of code given below R will spit
out the sum of squared error values for all of the parameters it is
trying. You should notice that the values decrease in size.</p>
<p>The last two lines of code in this set give you the parameter
estimates and an estimate of Ro. Here Ro is beta*S0/gamma. It includes
initial S (S0) because the ordinary differential equations we used here
did not include the population size. Getting the correct pop. size
estimate value for a specific variant and group of susceptible
individuals is hard, so we use a version of the equations without
population size.</p>
<pre class="r"><code># Initial parameter values for optimization
# alpha
parms0 &lt;- c(-9,-9,12,-2.6)

# Omicron
#parms0 &lt;- c(-9.5,-9,12.5,-2.6)

# Delta
#parms0 &lt;- c(-9,-6,12,-2.6)

# Beta
#parms0 &lt;- c(-10,-10,8,-2.6)

# optimization
fit1 &lt;- optim(parms0, sse.sir,data = pdata)</code></pre>
<pre><code>## [1] 1.503574e+12
## [1] 1.503599e+12
## [1] 1.477603e+12
## [1] 1.713482e+13
## [1] 1.503706e+12
## [1] 120495775920
## [1] 9297914966
## [1] 121638388854
## [1] 230772495270
## [1] 29932508649
## [1] 84794678795
## [1] 5132786006
## [1] 3024411625
## [1] 3367513173
## [1] 2869405689
## [1] 3759990197
## [1] 2843194222
## [1] 3544662709
## [1] 3150960965
## [1] 3807515247
## [1] 3360332895
## [1] 3229661201
## [1] 2866972099
## [1] 2932623505
## [1] 2808402936
## [1] 3150099287
## [1] 2895835217
## [1] 2880676280
## [1] 2835136172
## [1] 2760607173
## [1] 2719683069
## [1] 2754320257
## [1] 2730898805
## [1] 2823344493
## [1] 2762860151
## [1] 2691021610
## [1] 3280675097
## [1] 3058641182
## [1] 2747451067
## [1] 2862158951
## [1] 2719072496
## [1] 2801455340
## [1] 2719487222
## [1] 2662723101
## [1] 2622926983
## [1] 2664220568
## [1] 2667921163
## [1] 2596292654
## [1] 2539815083
## [1] 2608977953
## [1] 2578765247
## [1] 2527126028
## [1] 2495911301
## [1] 2428127495
## [1] 2657500325
## [1] 2259506873
## [1] 2154312147
## [1] 2030406127
## [1] 1955154597
## [1] 2131484384
## [1] 1980167044
## [1] 3500026337
## [1] 2227531633
## [1] 2288525653
## [1] 1985347903
## [1] 2470185471
## [1] 2034689526
## [1] 2371553750
## [1] 1953972312
## [1] 2006800616
## [1] 1949062865
## [1] 1956037578
## [1] 1933672801
## [1] 1898961217
## [1] 1958773148
## [1] 1987929122
## [1] 1919518553
## [1] 2016396645
## [1] 1917137903
## [1] 1919267038
## [1] 1907375072
## [1] 1881964510
## [1] 1877897470
## [1] 1911394888
## [1] 1892743335
## [1] 1895401114
## [1] 1884339090
## [1] 1941006040
## [1] 1888355257
## [1] 1864838616
## [1] 1850270025
## [1] 1858532282
## [1] 1863410335
## [1] 1866808816
## [1] 1862096710
## [1] 1833436395
## [1] 1809265709
## [1] 1807961121
## [1] 1778031758
## [1] 1789051850
## [1] 1803184026
## [1] 1764100302
## [1] 1741934475
## [1] 1688802662
## [1] 1609465440
## [1] 1692236314
## [1] 1701307911
## [1] 1620355314
## [1] 1656027342
## [1] 1577544560
## [1] 1535770937
## [1] 1449681781
## [1] 1322354258
## [1] 1302263078
## [1] 1136744179
## [1] 1200538577
## [1] 1270732194
## [1] 1263034888
## [1] 1222118063
## [1] 1068322190
## [1] 1317480267
## [1] 1084372644
## [1] 1080474163
## [1] 1115262715
## [1] 1089713723
## [1] 1014429568
## [1] 1048675337
## [1] 1185187152
## [1] 1074005911
## [1] 1021299579
## [1] 1032076149
## [1] 1005763235
## [1] 999317984
## [1] 1081669553
## [1] 1032488909
## [1] 946309800
## [1] 892222684
## [1] 945747123
## [1] 953667690
## [1] 977535244
## [1] 959741003
## [1] 887830511
## [1] 837997921
## [1] 808865895
## [1] 722864015
## [1] 722365488
## [1] 612047816
## [1] 642045753
## [1] 682375487
## [1] 584476889
## [1] 1104079102
## [1] 497935718
## [1] 554527296
## [1] 624246384
## [1] 545687997
## [1] 442777536
## [1] 375369756
## [1] 974992583
## [1] 524256323
## [1] 377806825
## [1] 420145759
## [1] 380498071
## [1] 396517795
## [1] 307608850
## [1] 266191552
## [1] 290594449
## [1] 285070153
## [1] 319345676
## [1] 297027308
## [1] 297037609
## [1] 283701131
## [1] 219423785
## [1] 220465132
## [1] 255750202
## [1] 246238615
## [1] 293331956
## [1] 257692174
## [1] 206214303
## [1] 180803604
## [1] 218883623
## [1] 214012420
## [1] 196813031
## [1] 200100475
## [1] 170521601
## [1] 166124104
## [1] 168696164
## [1] 169548492
## [1] 175827425
## [1] 166214223
## [1] 176100249
## [1] 162306009
## [1] 212530761
## [1] 164498394
## [1] 164067575
## [1] 162062288
## [1] 158952687
## [1] 158979137
## [1] 162486122
## [1] 160274666
## [1] 166034875
## [1] 160907009
## [1] 164511655
## [1] 160238222
## [1] 160939343
## [1] 159852965
## [1] 159794727
## [1] 159424086
## [1] 159350201
## [1] 159223950
## [1] 159155057
## [1] 159026446
## [1] 159170348
## [1] 158980205
## [1] 158619039
## [1] 158529108
## [1] 158689039
## [1] 158636860
## [1] 159161609
## [1] 158750639
## [1] 158866108
## [1] 158697059
## [1] 158283922
## [1] 158043074
## [1] 158481271
## [1] 158405464
## [1] 158022565
## [1] 157731210
## [1] 157970421
## [1] 158014412
## [1] 157586957
## [1] 157342226
## [1] 156974830
## [1] 156342628
## [1] 156766717
## [1] 156967771
## [1] 156025234
## [1] 155265006
## [1] 155830314
## [1] 155981048
## [1] 154574391
## [1] 153374414
## [1] 153964498
## [1] 154429036
## [1] 153314658
## [1] 152884602
## [1] 151139186
## [1] 148897256
## [1] 149777267
## [1] 150733015
## [1] 148408593
## [1] 146580140
## [1] 148679936
## [1] 148478365
## [1] 143583219
## [1] 139351393
## [1] 141478297
## [1] 143389271
## [1] 141673379
## [1] 142322058
## [1] 137356607
## [1] 135157986
## [1] 131170912
## [1] 125435541
## [1] 128189653
## [1] 130786739
## [1] 125590610
## [1] 127391276
## [1] 131525751
## [1] 126931393
## [1] 115296814
## [1] 104976843
## [1] 125674270
## [1] 120460995
## [1] 110867822
## [1] 114559728
## [1] 105302717
## [1] 109788884
## [1] 94404273
## [1] 80653700
## [1] 83776800
## [1] 90717913
## [1] 75471903
## [1] 66049497
## [1] 59771699
## [1] 51178450
## [1] 104728389
## [1] 64958567
## [1] 75241042
## [1] 59763128
## [1] 123659341
## [1] 64573116
## [1] 52293885
## [1] 54258524
## [1] 47402004
## [1] 45102392
## [1] 53255370
## [1] 47617563
## [1] 36844249
## [1] 33219529
## [1] 35659660
## [1] 36664791
## [1] 34315674
## [1] 33478151
## [1] 50088038
## [1] 37719846
## [1] 33445152
## [1] 32095919
## [1] 36473342
## [1] 33625025
## [1] 29546136
## [1] 27547989
## [1] 29944478
## [1] 29995975
## [1] 34770323
## [1] 30972704
## [1] 27265889
## [1] 26710086
## [1] 26064769
## [1] 28821311
## [1] 26818619
## [1] 26024358
## [1] 23939942
## [1] 25232284
## [1] 33002749
## [1] 25532233
## [1] 23830039
## [1] 23453942
## [1] 24213441
## [1] 23945368
## [1] 23467768
## [1] 23385856
## [1] 22721396
## [1] 23991536
## [1] 23157123
## [1] 22914692
## [1] 23931925
## [1] 23161167
## [1] 22839571
## [1] 22755875
## [1] 23082740
## [1] 22781237
## [1] 23603765
## [1] 22811607
## [1] 22732483
## [1] 22696672
## [1] 22956305
## [1] 22722344
## [1] 22717537
## [1] 22696893
## [1] 22646936
## [1] 22638678
## [1] 22729324
## [1] 22678715
## [1] 22711287
## [1] 22673040
## [1] 22686085
## [1] 22664743
## [1] 22609074
## [1] 22572537
## [1] 22644278
## [1] 22628302
## [1] 22627717
## [1] 22617130
## [1] 22582821
## [1] 22593458
## [1] 22559261
## [1] 22531798
## [1] 22517470
## [1] 22475467
## [1] 22476330
## [1] 22500701
## [1] 22450044
## [1] 22417563
## [1] 22396914
## [1] 22376638
## [1] 22344248
## [1] 22332776
## [1] 22283048
## [1] 22253662
## [1] 22417818
## [1] 22329645
## [1] 22455681
## [1] 22329818
## [1] 22298605
## [1] 22288201
## [1] 22275181
## [1] 22277540
## [1] 22329760
## [1] 22293081
## [1] 22211893
## [1] 22156786
## [1] 22203373
## [1] 22217183
## [1] 22148979
## [1] 22092843
## [1] 22083955
## [1] 22025637
## [1] 21958162
## [1] 21824495
## [1] 21810730
## [1] 21622394
## [1] 21610775
## [1] 21389378
## [1] 21410248
## [1] 21535051
## [1] 21073519
## [1] 20682977
## [1] 20747097
## [1] 20967639
## [1] 20739021
## [1] 20827248
## [1] 20582263
## [1] 20804769
## [1] 20292345
## [1] 20796847
## [1] 20549694
## [1] 20477148
## [1] 20236668
## [1] 20169134
## [1] 20853777
## [1] 20422514
## [1] 20010257
## [1] 19821628
## [1] 19760469
## [1] 19512991
## [1] 19908945
## [1] 19797707
## [1] 19793808
## [1] 19725342
## [1] 19392744
## [1] 19325962
## [1] 19341954
## [1] 19390038
## [1] 19116032
## [1] 19036034
## [1] 18954547
## [1] 19160694
## [1] 19511200
## [1] 19167885
## [1] 19117938
## [1] 19031231
## [1] 19008487
## [1] 18955921
## [1] 18836096
## [1] 18838186
## [1] 19015408
## [1] 18926910
## [1] 18766858
## [1] 18704075
## [1] 18718581
## [1] 18751072
## [1] 18695302
## [1] 18790337
## [1] 18792916
## [1] 18714375
## [1] 18546578
## [1] 18453065
## [1] 18760451
## [1] 18638418
## [1] 18529171
## [1] 18556698
## [1] 18426948
## [1] 18331373
## [1] 18421711
## [1] 18421123
## [1] 18237232
## [1] 18111089
## [1] 18201572
## [1] 18234739
## [1] 18111779
## [1] 18160859
## [1] 17982900
## [1] 17902282
## [1] 18035007
## [1] 18001620
## [1] 17833477
## [1] 17745548
## [1] 17764620
## [1] 17810858
## [1] 17910741
## [1] 17817787
## [1] 17630162
## [1] 17565859
## [1] 17720254
## [1] 17656123
## [1] 17532628
## [1] 17530900
## [1] 17680789
## [1] 17580502
## [1] 17557693
## [1] 17516939
## [1] 17445150
## [1] 17491445
## [1] 17363100
## [1] 17273499
## [1] 17654570
## [1] 17454494
## [1] 17456217
## [1] 17410286
## [1] 17249620
## [1] 17131716</code></pre>
<pre class="r"><code># parameter values (beta, gamma, S0, I0)
exp(fit1$par)</code></pre>
<pre><code>## [1] 7.296273e-06 4.887612e-01 1.420334e+05 1.623186e-01</code></pre>
<pre class="r"><code># Estimate of Ro
exp(fit1$par[1])*exp(fit1$par[3])/exp(fit1$par[2])</code></pre>
<pre><code>## [1] 2.120288</code></pre>
<p>The next code block produces a plot of the observed and optimized
model values for the number of infected individuals.</p>
<pre class="r"><code>#produce model predicted values to plot
mod.pred &lt;- data.frame(ode(c(exp(fit1$par[3]),exp(fit1$par[4])),t,
                           sir,c(exp(fit1$par[1]),exp(fit1$par[2])),hmax=1/120))
names(mod.pred) &lt;- c(&quot;t&quot;,&quot;S&quot;,&quot;I&quot;)

#plot observed and modeled predictions
(Iplot &lt;- ggplot(pdata, aes (x = t, y = I)) +
  geom_point(alpha = 0.5, size = 3) + 
  geom_line(data = mod.pred, aes(x = t, y = I), col = &quot;red&quot;, size = 1.5) + 
  theme_bw() +
  ylab(&quot;Infected\n&quot;) +                             
  xlab(&quot;\nWeek&quot;)  +
  theme(axis.text.x = element_text(size = 14),
        axis.text.y = element_text(size = 14),
        axis.title = element_text(size = 16, face = &quot;plain&quot;),                        
        panel.grid = element_blank(),                                                 
        plot.margin = unit(c(1,1,1,1), units = , &quot;cm&quot;)))</code></pre>
<p><img src="DSEE-lab7_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p><mark style="color:blue"><em>After you work through all of the above
code for each of the four variants, I want you to alter the code so that
you can produce a SINGLE publication quality plot that has the observed
and optimized model infection values for all four
variants.</em></mark></p>
<p>Do the Ro values for any of the variants seem too high to low? If you
are unsure, Google the current best estimates of the Ro values for each
of the variants. What could be possible causes for these inaccurate
estimates? Think both about issues with optimization (hint: look at the
plots) and violation of SIR assumptions.</p>
<p><br></p>
</div>
<div id="classify-covid-19-variants-and-predict-ro-for-emerging-variant"
class="section level3">
<h3>2. Classify covid-19 variants and predict Ro for emerging
variant</h3>
<p>Next up, we are going to use sars-cov-2 spike protein sequences to
classify variants and, in conjunction with results from part 1, estimate
the Ro for an emerging variant. The spike protein plays an important
role in binding of the sars-cov-2 virus to human cells, and influences
the transmission and virulence of the disease (eg, <a
href="https://www.nature.com/articles/s41401-020-0485-4">Huang et al
2020</a>). The protein sequences come from NCBI, which has a decent <a
href="https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/">dashboard</a> for
downloading covid related data.</p>
<p>To classify the emerging variant, we are going to use <a
href="https://en.wikipedia.org/wiki/Principal_component_analysis">principle
component analysis</a> (PCA) to reduce the dimensionality of the data
(protein sequences are quite complex), then use the k-nearest neighbor
method (remember back to the lecture from lab last week) to determine
which of the <a
href="https://www.who.int/en/activities/tracking-SARS-CoV-2-variants/">variants
of concern</a> the emerging variant is most similar to and use the
variant of concern Ro as a first step prediction of the emerging
variant’s basic reproductive number.</p>
<p>The k-nearest neighbor method uses a training dataset (called
sequences_train on Canvas), in this case, one where we know the variants
(and their Ro values). We use the training dataset to classify the
emerging, or unknown, variant (found in the test dataset called
sequences_test on Canvas). The training dataset is composed of 20
sequences for each of four variants of concern: Alpha, Beta, Delta, and
Omicron. The test dataset is composed of 5 sequences.</p>
<p>Let’s start by getting the needed packages and data into R.</p>
<pre class="r"><code># required packages
require(ape)
require(kmer)
require(class)
require(ggfortify)

# training data, the data come in what is called fasta format
train &lt;- read.FASTA(&quot;/Users/17163/Documents/DSEE/Labs/Covid/sequences_train.txt&quot;,type=&quot;AA&quot;)

# training data IDs (variant)
vID &lt;- c(rep(&quot;Alpha&quot;,10),rep(&quot;Delta&quot;,10),rep(&quot;Beta&quot;,10),rep(&quot;Omicron&quot;,10),
               rep(&quot;Alpha&quot;,10),rep(&quot;Beta&quot;,10),rep(&quot;Delta&quot;,10),rep(&quot;Omicron&quot;,10))

# test data (unknowns)
test &lt;- read.FASTA(&quot;/Users/17163/Documents/DSEE/Labs/Covid/sequences_test.txt&quot;,type=&quot;AA&quot;)

# combining training and test data
comb &lt;- c(train,test)

# adding test data unknown variant
vID_t &lt;- c(vID, rep(&quot;U&quot;,5))</code></pre>
<p>The first step in our classification procedure is to produce a
distance matrix. This is basically a measure of similarity between the
sequences. The higher the number the more similar the sequences. But
then we have an 85x85 matrix of distance values. So, we use principal
component analysis to reduce the complexity down to two variables using
PCA. PCA maximizes the amount of variation that can be explained in a
smaller number of variables, which in this case is two. PCA also allows
us to plot the data more easily.</p>
<pre class="r"><code># calculate distance matrix
mat_t &lt;- kdistance(comb,k=2, method = &quot;edgar&quot;)

# reformat and calculate principle components
out_t &lt;- kmeans(mat_t,4)
re_mat_t &lt;- as.matrix(mat_t)
plot &lt;- autoplot(out_t, mat_t, frame = TRUE)</code></pre>
<p>Before we use these data to help classify the unknown emerging
variant sequences, let’s plot the results of the PCA.</p>
<pre class="r"><code># make data frame of PCs and Variant IDs
plot_data &lt;- data.frame(plot$data$PC1,plot$data$PC2,vID_t)
names(plot_data) &lt;- c(&quot;dem1&quot;,&quot;dem2&quot;,&quot;Variant&quot;)

# colors for plot
colors &lt;- c(&quot;red&quot;,&quot;green&quot;,&quot;blue&quot;,&quot;purple&quot;,&quot;orange&quot;)

# plot data
(d_plot &lt;- ggplot(plot_data, aes (x = dem1, y = dem2, color=Variant)) +
    geom_point(alpha = 0.5, size = 3) + 
    scale_color_manual(values = colors) +
    theme_bw() +
    ylab(&quot;PC2\n&quot;) +                             
    xlab(&quot;\nPC1&quot;)  +
    theme(axis.text.x = element_text(size = 14),
          axis.text.y = element_text(size = 14),
          axis.title = element_text(size = 16, face = &quot;plain&quot;),                        
          panel.grid = element_blank(),                                                 
          plot.margin = unit(c(1,1,1,1), units = , &quot;cm&quot;)))</code></pre>
<p><img src="DSEE-lab7_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<p>The plot has the principal component variables 1 and 2 on the x and y
axes and variants IDs by color. You can see that the PCs do a pretty
good (not perfect!) job of grouping the sequences for variants together.
For example, all of the Omicron sequences are clumped nicely at the top
in purple.</p>
<p>The last step is to use the k-nearest neighbor method to determine
which variant of concern the uknowns are most similar to. More
specifically, the code below finds the five nearest neighbors for each
of the unknowns and outputs the variant it would assign each emerging
sequence to. It outputs them simply as a list.</p>
<pre class="r"><code># finds the 20 (k=20) nearest neighbors for the unknowns and assigns  
# them to a group (variant)
knn(plot_data[1:80,1:2],plot_data[81:85,1:2],cl=plot_data[1:80,3],k=20, prob = T)</code></pre>
<pre><code>## [1] Omicron Omicron Omicron Omicron Omicron
## attr(,&quot;prob&quot;)
## [1] 0.9 0.9 0.9 0.9 0.9
## Levels: Alpha Beta Delta Omicron</code></pre>
<p>You may have also noticed that I specified an argument called “prob”
to be true in the above code. This give the proportion of nearest
neighbors within the class or group to which the algorithm assigns the
unknowns (emerging variant sequences).</p>
<p>I developed a function (given below) that uses this proportion of
nearest neighbors to estimate the Ro with a weighted mean based on the
proportions given.</p>
<pre class="r"><code># function for finding k nearest neighbors
# and calculating the proportion for each class (variant)

# make data frame of estimated Ro by variant (order is Alpha, Delta, Beta, Omicron)
Ro &lt;- c(1.8,5.8,1,8.5)
Variant &lt;- unique(vID_t)[-length(unique(vID_t))]
Ro_d &lt;- data.frame(Ro,Variant)

# x is the PCs, y is vID_t (variant labels), z is the number of neighbors considered, zz is Ro_d(Ro for each variant)
knnd &lt;- function(x,y,z,zz) {
  d1 &lt;- as.matrix(dist(x[,1:ncol(x)])) # generate distance matrix
  d2 &lt;- d1[y==&quot;U&quot;,y!=&quot;U&quot;]              # subset only unknowns
  o &lt;- y[apply(d2,1,order)[1:z,]]      # order by nearest neighbors
  o &lt;- matrix(o,nrow=z)                # convert to matrix
  tt &lt;- apply(o,2,table)               # determine proportion of z neighbors for each variant
  Ro_w&lt;-numeric(ncol(o))
  var&lt;-numeric(ncol(o))
  for (i in 1:ncol(o)){
    Ro_w[i]&lt;-sum(Ro_d$Ro[match(names(tt[[i]]),Ro_d$Variant)]*tt[[i]])/z #calc weighted mean
    if (length(names(which.max(tt[[i]])))==0) {var[i]&lt;-NA} else {var[i]&lt;-names(which.max(tt[[i]]))} # determine most similar variant
  }
  print(var)  # print results
  print(Ro_w) 
}

# run function on your data
knnd(plot_data[1:85,1:2],vID_t,20,Ro_d)</code></pre>
<pre><code>## [1] NA NA NA NA NA
## [1] 0 0 0 0 0</code></pre>
<p><mark style="color:blue"><em>Write these results in your own words
and be sure to include your best estimate for the Ro.</em></mark></p>
<p><mark style="color:blue"><em>Together in lab, we will go through the
process of finding and downloading a sequence from the NCBI website.
Then, you can repeat the process in part 2 to make predictions about
another “unknown variant” sequence.</em></mark></p>
<p><mark style="color:blue"><em>Be sure to save the principal component
plot with your additional sequence plotted and an interpretation of your
results.</em></mark></p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
