<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Data Visualization - Part 1</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">DSEE Labs</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="DSEE-Lab1.html">DSEE Lab #1</a>
</li>
<li>
  <a href="DSEE-Lab2.html">DSEE Lab #2</a>
</li>
<li>
  <a href="DSEE-Lab3.html">DSEE Lab #3</a>
</li>
<li>
  <a href="DSEE-Lab4.html">DSEE Lab #4</a>
</li>
<li>
  <a href="DSEE-Lab5and6.html">DSEE Lab #5 and 6</a>
</li>
<li>
  <a href="DSEE-Lab7.html">DSEE Lab #7</a>
</li>
<li>
  <a href="DSEE-Lab8.html">DSEE Lab #8</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Data Visualization - Part 1</h1>

</div>


<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('left', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<p><em>Learning goals and steps:</em></p>
<p><em>1. Get familiar with the ggplot2 syntax</em></p>
<p><em>2. Decide on the right type of plot</em></p>
<p><em>3. Practice making different plots with ggplot2</em></p>
<div class="line-block">    <em>-Histograms</em><br />
    <em>-Scatter plots</em><br />
    <em>-Box plots</em></div>
<p><br></p>
<p><strong>Due:</strong> Nothing this week</p>
<p>For the next two labs we will focus on communicating results using
visualizations, or graphics. The data we are going to use for these labs
is from a paper published in 2013 on the role of biodiversity in
mitigating phenological mismatches. The <a
href="https://fandm.instructure.com/courses/17827/files?preview=1433579">paper</a>
uses a long term dataset of bloom times for apple trees (<em>Malus</em>
ssp.) and flight times for apple flower pollinators.</p>
<p>For part 1, there are a few bits of code and several plots that you
will need to save for your write-up. They are indicated with the
<em><mark style="color:blue">blue text</mark></em>.</p>
<p><br></p>
<div id="good-data-visualization-and-ggplot2-syntax"
class="section level3">
<h3>1. Good data visualization and ggplot2 syntax</h3>
<p>When it comes to data visualization, the package ggplot2 by Hadley
Wickham has won over many scientists’ hearts. In this tutorial, we will
learn how to make beautiful and informative graphs and how to arrange
them in a panel. Before we tackle the ggplot2 syntax, let’s briefly
cover what good graphs have in common.</p>
<p><img src="images/Plot%20table.PNG" /></p>
<p>ggplot2 is a great package to guide you through those steps. The gg
in ggplot2 stands for grammar of graphics. Writing the code for your
graph is like constructing a sentence made up of different parts that
logically follow from one another. In a more visual way, it means adding
layers that take care of different elements of the plot. Your plotting
workflow will therefore be something like creating an empty plot, adding
a layer with your data points, then your measure of uncertainty, the
axis labels, and so on.</p>
<p><img src="images/Graphs%20in%20ggplot2.PNG" /></p>
<p><br></p>
</div>
<div id="decide-on-the-right-type-of-plot" class="section level3">
<h3>2. Decide on the right type of plot</h3>
<p>A very key part of making any data visualization is making sure that
it is appropriate to your data type (e.g. discrete vs continuous), and
fits your purpose, i.e. what you are trying to communicate!</p>
<p>You can start with our simple guide for common graph types, and visit
the <a href="https://r-graph-gallery.com/">R Graph Gallery</a>, a
fantastic resource for ggplot2 code and inspiration!</p>
<p><img src="images/Graphs%20in%20ggplot2%202.PNG" /></p>
<p><br></p>
</div>
<div id="making-different-plots-with-ggplot2" class="section level3">
<h3>3. Making different plots with ggplot2</h3>
<p>Before we make plots, let’s set up the script. <strong>Do not forget
to annotate!!!</strong></p>
<pre class="r"><code># Purpose of the script
# Your name, date and email

# Libraries - if you haven&#39;t installed them before, run the code install.packages(&quot;package_name&quot;)
library(tidyr)
library(dplyr)
library(ggplot2)
library(readr)
library(gridExtra)</code></pre>
<p>Next, we need to get data into R. There are three datasets provided
for you on Canvas:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Mid_bloom.csv</strong> – mean mid-bloom date of apple
trees by year</p></li>
<li><p><strong>Pollinators.csv</strong> – data on pollinators including
scientific name, date collected (Julian day of year), year collected,
and corrected collection day (controlling for effect of latitude). You
should use the corrected collection day and not collection
date.</p></li>
<li><p><strong>Climate.csv</strong> – Has daily maximum and minimum
temperature (along with other variables) for the field site from 1969
through 2010.</p></li>
</ol>
<p>Get them uploaded into R (review “Introduction to R” labs for
assistance).</p>
<p><br></p>
</div>
<div id="a.-histograms-to-visualize-data-distribution"
class="section level3">
<h3>3a. Histograms to visualize data distribution</h3>
<p>We are going to make a histogram of asynchrony from the Pollinator
dataset, but first we need to calculate it. Asynchrony is the difference
between mid-bloom date of the apples and the collection date for the
pollinators. Collection date is an estimate of the flight period (when
pollinators are out visiting plants). The bigger the asynchrony the more
phenological mismatch.</p>
<pre class="r"><code>#calculate asynchrony

#join mid bloom date with pollinators
Pollinators &lt;- left_join(Pollinators, Mid_bloom, by = c(&quot;Year&quot;))

#calculate asynchrony and add to the dataset using mutate
Pollinators &lt;- Pollinators %&gt;%
  mutate(Asynchrony = CorrectedCollectionDay - Mid_bloom_day)</code></pre>
<p>Then, we need to filter the data so that we only keep pollinators
with absolute asynchrony values less than or equal to 25 (hint:
<mark style="background-color:grey">abs()</mark> is a function that
produces absolute value). We do this to avoid including pollinators that
do not normally forage when the apple trees are blooming.</p>
<p><mark style="color:blue"><em>Write the code to filter the data and
include it in your write-up.</em></mark></p>
<p>We will do a quick comparison between base R graphics and ggplot2 -
of course both can make good graphs when used well, but we like working
with ggplot2 because of its powerful customization abilities.</p>
<pre class="r"><code># With base R graphics
base_hist &lt;- hist(Pollinators$Asynchrony)</code></pre>
<p><img src="DSEE-Lab3_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>Using <mark style="background-color:grey">gglot()</mark>, we specify
the type of graph using
<mark style="background-color:grey">geom_histogram()</mark>. Note that
putting your entire ggplot code in brackets () creates the graph and
then shows it in the plot viewer. If you don’t have the brackets, you’ve
only created the object, but haven’t visualized it. You would then have
to call the object in the command line, e.g. by typing
<mark style="background-color:grey">asynchrony_hist</mark> after
creating the object.</p>
<pre class="r"><code># With ggplot2: creating graph with no brackets
asynchrony_hist &lt;- ggplot(Pollinators, aes(x = Asynchrony))  +
  geom_histogram() 
  
# Calling the object to display it in the plot viewer
asynchrony_hist</code></pre>
<p><img src="DSEE-Lab3_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<pre class="r"><code># With brackets: you create and display the graph at the same time
(asynchrony_hist &lt;- ggplot(Pollinators, aes(x = Asynchrony))  +
  geom_histogram())</code></pre>
<p><img src="DSEE-Lab3_files/figure-html/unnamed-chunk-6-2.png" width="672" /></p>
<p>The default <mark style="background-color:grey">ggplot()</mark>
settings are not ideal: there is lots of unnecessary gray space behind
the histogram, the axis labels are quite small, and the bars blend with
each other. Lets beautify the histogram a bit! This is where the true
power of ggplot2 shines.</p>
<pre class="r"><code>(asynchrony_hist &lt;- ggplot(Pollinators, aes(x = Asynchrony)) +                
  geom_histogram(binwidth = 2, color = &quot;#8B5A00&quot;, fill = &quot;#CD8500&quot;) +    # Changing the binwidth and colors
  geom_vline(aes(xintercept = mean(Asynchrony)),                       # Adding a line for mean asynchrony
             color = &quot;red&quot;, linetype = &quot;dashed&quot;, size=1) +           # Changing the look of the line
    theme_bw() +                                                      # Changing the theme to get rid of the gray background
  ylab(&quot;Frequency\n&quot;) +                                                   # Changing the text of the y axis label
  xlab(&quot;\nPollinator asynchrony&quot;)  +                              # \n adds a blank line between axis and text
  theme(axis.title.x = element_text(size = 14, face = &quot;plain&quot;),   # x axis face=&quot;plain&quot; is the default, you can change it to italic, bold, etc. 
        axis.title.y = element_text(size = 14, face = &quot;plain&quot;),       # y axis 
        panel.grid = element_blank(),                                 # Removing the gray grid lines
        plot.margin = unit(c(1,1,1,1), units = , &quot;cm&quot;)))              # Putting a 1 cm margin around the plot</code></pre>
<p><img src="DSEE-Lab3_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p><mark style="color:blue"><em>Save this plot for your Data
visualization write-up.</em></mark></p>
<p>Note: Pressing enter after each “layer” of your plot (i.e. indenting
it) prevents the code from being one gigantic line and makes it much
easier to read.</p>
<p><br></p>
</div>
<div id="understanding-ggplot2s-jargon" class="section level3">
<h3><mark style="background-color:lightblue">Understanding ggplot2’s
jargon</mark></h3>
<p><mark style="background-color:lightblue">Perhaps the trickiest bit
when starting out with ggplot2 is understanding what type of elements
are responsible for the contents (data) versus the container (general
look) of your plot. Let’s de-mystify some of the common words you will
encounter.</mark></p>
<p><mark style="background-color:lightblue">geom: a geometric object
which defines the type of graph you are making. It reads your data in
the aesthetics mapping to know which variables to use, and creates the
graph accordingly. Some common types are
</mark><mark style="background-color:grey">geom_point()</mark><mark style="background-color:lightblue">,
</mark><mark style="background-color:grey">geom_boxplot()</mark><mark style="background-color:lightblue">,
</mark><mark style="background-color:grey">geom_histogram()</mark><mark style="background-color:lightblue">,
</mark><mark style="background-color:grey">geom_col()</mark><mark style="background-color:lightblue">,
etc.</mark></p>
<p><mark style="background-color:lightblue">aes: short for aesthetics.
Usually placed within a
</mark><mark style="background-color:grey">geom_</mark><mark style="background-color:lightblue">,
this is where you specify your data source and variables, AND the
properties of the graph which depend on those variables. For instance,
if you want all data points to be the same color, you would define the
<mark style="background-color:grey">color
=</mark><mark style="background-color:lightblue"> argument outside the
</mark><mark style="background-color:grey">aes()</mark><mark style="background-color:lightblue">
function; if you want the data points to be colored by a factor’s levels
(e.g. by site or species), you specify the
</mark><mark style="background-color:grey">color
=</mark><mark style="background-color:lightblue"> argument inside the
</mark><mark style="background-color:grey">aes()</mark><mark style="background-color:lightblue">.</mark></p>
<p><mark style="background-color:lightblue">stat: a stat layer applies
some statistical transformation to the underlying data: for instance,
</mark><mark style="background-color:grey">stat_smooth(method =
‘lm’)</mark><mark style="background-color:lightblue"> displays a linear
regression line and confidence interval ribbon on top of a scatter plot
(defined with
</mark><mark style="background-color:grey">geom_point()</mark><mark style="background-color:lightblue">).</mark></p>
<p><mark style="background-color:lightblue">theme: a theme is made of a
set of visual parameters that control the background, borders, grid
lines, axes, text size, legend position, etc. You can use <a
href="https://ourcodingclub.github.io/tutorials/data-vis-2/index.html#theme">pre-defined
themes, create your own</a>, or use a theme and overwrite only the
elements you don’t like. Examples of elements within themes are
</mark><mark style="background-color:grey">axis.text</mark><mark style="background-color:lightblue">,
</mark><mark style="background-color:grey">panel.grid</mark><mark style="background-color:lightblue">,
</mark><mark style="background-color:grey">legend.title</mark><mark style="background-color:lightblue">,
and so on. You define their properties with
</mark><mark style="background-color:grey">elements_…()</mark><mark style="background-color:lightblue">
functions:
</mark><mark style="background-color:grey">element_blank()</mark><mark style="background-color:lightblue">
would return something empty (ideal for removing background color),
while </mark><mark style="background-color:grey">element_text(size = …,
face = …, angle = …)</mark><mark style="background-color:lightblue">
lets you control all kinds of text properties.</mark></p>
<p><mark style="background-color:lightblue">It is also useful to
remember] that layers are added on top of each other as you progress
into the code, which means that elements written later may hide or
overwrite previous elements.</mark></p>
<p><br></p>
</div>
<div id="learning-how-to-use-colourpicker" class="section level3">
<h3>Learning how to use colourpicker</h3>
<p>In the code above, you can see a color code
<mark style="background-color:grey">colour = “#8B5A00”</mark> - any
color you can dream of has a code, called a “hex code”, a combination of
letters and numbers. You can get the codes for different colors online,
from Paint, Photoshop or similar programs, or even from RStudio, which
is very convenient! There is an RStudio
<mark style="background-color:grey">Colourpicker</mark> addin which was
a game changer for us - to install it, run the following code:</p>
<pre class="r"><code># package for generating color hex codes
install.packages(&quot;colourpicker&quot;)</code></pre>
<p>To find out the code for a color you like, click on
<mark style="background-color:grey">Addins/Colour picker</mark>.</p>
<p><img src="images/Colourpicker.PNG" /></p>
<p>When you click on <mark style="background-color:grey">All R
colours</mark> (bottom right) you will see lots of different colors you
can choose from - a good color scheme makes your graph stand out, but of
course, don’t go crazy with the colors. When you click on
<mark style="background-color:grey">1</mark>, and then on a certain
color, you fill up <mark style="background-color:grey">1</mark> with
that color, same goes for <mark style="background-color:grey">2</mark>,
<mark style="background-color:grey">3</mark> - you can add more colors
with the <mark style="background-color:grey">+</mark>, or delete them by
clicking the bin. Once you’ve made your pick, click
<mark style="background-color:grey">Done</mark>. You will see a line of
code <mark style="background-color:grey">c(“#8B5A00”, “#CD8500”)</mark>
appear in your script - in this case, we just need the color code, so we
can copy that, and delete the rest. Try changing the color of the
histogram you made just now.</p>
<p><img src="images/Colourpicker2.PNG" /></p>
<p><br></p>
</div>
<div id="b.-scatter-plot-to-examine-change-over-time"
class="section level3">
<h3>3b. Scatter plot to examine change over time</h3>
<p>We have taken a quick look at asynchrony, and we will come back to
it. But let’s examine some of the underlying data by making a
scatterplot of pollinator collection date by time and family.</p>
<pre class="r"><code># Make family a factor (categorical variable)
Pollinators$Family &lt;- as.factor(Pollinators$Family)

# Using default ggplot2 graphics
(pollinator_scatter &lt;- ggplot(Pollinators, aes(x = Year, y = CorrectedCollectionDay, color = Family)) +  
                             geom_point())</code></pre>
<p><img src="DSEE-Lab3_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<p>Hopefully by now we’ve convinced you of the perks of ggplot2, but
again like with the histogram, the graph above needs a bit more work. We
are adding best fit lines, along with other features.</p>
<pre class="r"><code># Create vector colors to be used in plot for each family
colors &lt;-  c(&quot;#0000EE&quot;, &quot;#8A2BE2&quot;, &quot;#FF1493&quot;, &quot;#CDAD00&quot;, &quot;#548B54&quot;, &quot;#8B5A00&quot;)

(pollinator_scatter &lt;- ggplot(Pollinators, aes (x = Year, y = CorrectedCollectionDay, color = Family)) +
    geom_point(size = 2) +                                               # Changing point size
    geom_smooth(method = &quot;lm&quot;, aes(fill = Family)) +               # Adding linear model fit, color-code by country
    theme_bw() +
    scale_fill_manual(values = colors) +                # Adding custom colors for solid geoms (ribbon)
    scale_color_manual(values = colors) +              # Adding custom colors for lines and points
    ylab(&quot;Collection Day\n&quot;) +                             
    xlab(&quot;\nYear&quot;)  +
    theme(axis.text.x = element_text(size = 12, angle = 45, vjust = 1, hjust = 1),     # making the years at a bit of an angle
          axis.text.y = element_text(size = 12),
          axis.title = element_text(size = 12, face = &quot;plain&quot;),                        
          panel.grid = element_blank(),                                   # Removing the background grid lines               
          plot.margin = unit(c(1,1,1,1), units = , &quot;cm&quot;),                 # Adding a 1cm margin around the plot
          legend.text = element_text(size = 12, face = &quot;italic&quot;),         # Setting the font for the legend text
          legend.title = element_blank(),                                 # Removing the legend title
          legend.position = c(0.87, 0.84)))                                 # Setting legend position - 0 is left/bottom, 1 is top/right</code></pre>
<p><img src="DSEE-Lab3_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<p>That’s cluttered! Can you really figure out what the families are
doing? By adding a facetting layer, we can split the data in multiple
facets representing the different families. This is done using
<mark style="background-color:grey">facet_wrap()</mark>.</p>
<pre class="r"><code># Create colors to be used in plot
colors &lt;-  c(&quot;#0000EE&quot;, &quot;#8A2BE2&quot;, &quot;#FF1493&quot;, &quot;#CDAD00&quot;, &quot;#548B54&quot;, &quot;#8B5A00&quot;)

# Plot the collection day by time for countries individually
(pollinator_scatter &lt;- ggplot(Pollinators, aes (x = Year, y = CorrectedCollectionDay, color = Family)) +
    geom_point(size = 2) +                              # Changing point size
    geom_smooth(method = &quot;lm&quot;, aes(fill = Family)) +    # Adding linear model fit, color-code by country
    theme_bw() +
    facet_wrap(~ Family, scales = &quot;free_y&quot;) +           # THIS LINE CREATES THE FACETTING
    scale_fill_manual(values = colors) +                # Adding custom colors for solid geoms (ribbon)
    scale_color_manual(values = colors) +               # Adding custom colors for lines and points
    ylab(&quot;Collection Day\n&quot;) +                             
    xlab(&quot;\nYear&quot;)  +
    theme(axis.text.x = element_text(size = 12, angle = 45, vjust = 1, hjust = 1),     # making the years at a bit of an angle
          axis.text.y = element_text(size = 12),
          axis.title = element_text(size = 12, face = &quot;plain&quot;),                        
          panel.grid = element_blank(),                 # Removing the background grid lines               
          plot.margin = unit(c(1,1,1,1), units = , &quot;cm&quot;),   # Adding a 1cm margin around the plot
          legend.text = element_text(size = 12, face = &quot;italic&quot;),   # Setting the font for the legend text
          legend.title = element_blank(),   # Removing the legend title
          legend.position = &quot;right&quot;))       # Setting legend position</code></pre>
<p><img src="DSEE-Lab3_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<p>Some useful arguments to include in
<mark style="background-color:grey">facet_wrap()</mark> are
<mark style="background-color:grey">nrow =</mark> or
<mark style="background-color:grey">ncol =</mark> , specifying the
number of rows or columns, respectively. You can also see that we used
<mark style="background-color:grey">scales = “free_y”</mark>, to allow
different y axis values because of the wide range of values in the data.
You can use “fixed” when you want to constrain all axis values.</p>
<p><br></p>
<div id="good-to-know" class="section level4">
<h4><mark style="background-color:#F9E59B">Good to know</mark></h4>
<p><mark style="background-color:#F9E59B">If your axis labels need to
contain special characters or superscript, you can get ggplot2 to plot
that, too. It might require some googling regarding your specific case,
but for example, this code
</mark><mark style="background-color:grey">ylabs(expression(paste(‘Grain
yield’,’ ‘,’(ton.’, ha^-1, ‘)’,
sep=’’)))</mark><mark style="background-color:#F9E59B"> will create a y
axis with a label reading Grain yield (ton. ha-1).</mark></p>
<p><br></p>
</div>
</div>
<div
id="c.-boxplot-to-examine-whether-asynchrony-differs-between-families"
class="section level3">
<h3>3c. Boxplot to examine whether asynchrony differs between
families</h3>
<p>Box plots are very informative as they show the median and spread of
your data, and allow you to quickly compare values among groups.</p>
<pre class="r"><code># boxplot of pollinator family asynchrony
(pollinator_boxplot &lt;- ggplot(Pollinators, aes(Family, Asynchrony)) + geom_boxplot())</code></pre>
<p><img src="DSEE-Lab3_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<pre class="r"><code># Beautifying
(pollinator_boxplot &lt;- ggplot(Pollinators, aes(Family, Asynchrony)) + 
    geom_boxplot(aes(fill = Family)) +
    theme_bw() +
    scale_fill_manual(values = colors) +      # Adding custom colors
    scale_color_manual(values = colors) +     # Adding custom colors
    ylab(&quot;Asynchrony\n&quot;) +                             
    xlab(&quot;\nFamily&quot;)  +
    theme(axis.text = element_text(size = 12),
          axis.title = element_text(size = 12, face = &quot;plain&quot;),                     
          panel.grid = element_blank(),                     # Removing the background grid lines               
          plot.margin = unit(c(1,1,1,1), units = , &quot;cm&quot;),   # Adding a margin
          legend.position = &quot;none&quot;))                        # Removing legend</code></pre>
<p><img src="DSEE-Lab3_files/figure-html/unnamed-chunk-12-2.png" width="672" /></p>
<p>We have made several plots today! Do any patterns you saw in the
plots tell us something interesting about the relationships between
plants, pollinators, and climate?</p>
<p><br></p>
</div>
<div id="challenge-yourself" class="section level3">
<h3>4. Challenge yourself!</h3>
<p><mark style="color:blue"><em>Using your new found graphing powers, I
want you to make and save the following plots for your
write-up:</em></mark></p>
<div class="line-block">      <mark style="color:blue"><em>1) A fully
formatted (aka beautified) plot of apple mid-bloom date by
year.</em></mark></div>
<div class="line-block">      <mark style="color:blue"><em>2) A fully
formatted plot of corrected collection day by pollinator
genera.</em></mark></div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
